src/Microsoft.Diagnostics.Runtime/Native/NativeHeap.cs

```diff
@@ -84,6 +154,121 @@ private void CreateFreeType()
             _types.Add(_free);
         }
 
+        public override ClrType GetClrTypeFromEE(ulong eeType)
+        {
+            if ((((int)eeType) & 3) != 0)
+                eeType &= ~3UL;
+
+            ClrType clrType = null;
+            int index;
+            if (_indices.TryGetValue(eeType, out index))
+                clrType = _types[index];
+            else
+                clrType = ConstructObjectType(eeType);
+
+            return clrType;
+        }
+
+        internal class NativeException : ClrException
+        {
+            private Address _ccwPtr;
+            private ClrType _type;
+            private ulong _hResult;
+            private ulong _threadId;
+            private ulong _exceptionId;
+            private ulong _innerExceptionId;
+            private ulong _nestingLevel;
+            private IList<ClrStackFrame> _stackFrames;
+
+            private ClrException _innerException;
+            
+            public NativeException(
+                            ClrType clrType,
+                            Address ccwPtr,
+                            ulong hResult,
+                            ulong threadId,
+                            ulong exceptionId,
+                            ulong innerExceptionId,
+                            ulong nestingLevel,
+                            IList<ClrStackFrame> stackFrames)
+            {
+                _type = clrType;
+                _ccwPtr = ccwPtr;
+                _hResult = hResult;
+                _threadId = threadId;
+                _exceptionId = exceptionId;
+                _innerExceptionId = innerExceptionId;
+                _nestingLevel = nestingLevel;
+                _stackFrames = stackFrames;
+            }
+
+            public override Address Address
+            {
+                get { return _ccwPtr; }
+            }
+            public override int HResult
+            {
+                get
+                {
+                    return (int)_hResult;
+                }
+            }
+
+            public override ClrType Type
+            {
+                get { return _type; }
+            }
+
+            public override string Message
+            {
+                get
+                {
+                    return "<Invalid Object>"; // we do not expect a Message in serialized exceptions, however, returning <Invalid Object> as the SOS does
```

Comment:
Better to `return null;` here.  This matches the rest of what ClrMD does.  It tells the user we do not have this message, and they can decide what to show ("<Invalid Object>" or something of their own choosing).  If we return a special string here, consumers of the API will have to handle this case.